### ISSUE TITLE
Tracking Issue for File lock API

### ISSUE THREAD

**Issue by cberner** [rust-lang/rust#130994]
Feature gate: `#![feature(file_lock)]`

This is a tracking issue for https://github.com/rust-lang/libs-team/issues/412

This feature exposes advisory file locks on `File`. They allow a file handle to acquire an exclusive or shared file lock, which blocks other file handles to the same file from acquiring a conflicting lock. Some semantics are platform dependent, and these are documented in the API documentation.

### Public API

```rust
impl File {
    fn lock(&self) -> io::Result<()>;
    fn lock_shared(&self) -> io::Result<()>;
    fn try_lock(&self) -> io::Result<bool>;
    fn try_lock_shared(&self) -> io::Result<bool>;
    fn unlock(&self) -> io::Result<()>;
}
```

### Steps / History

- [x] Implementation: #130999 
- [x] Final comment period (FCP)[^1]
- [x] Stabilization PR
  - https://github.com/rust-lang/rust/pull/136794

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by workingjubilee** [comment]
apparently not supported on all tier 2 OS: https://github.com/rust-lang/rust/issues/132921

**Comment by eric-seppanen** [comment]
Note that this triggers the `unstable_name_collisions` lint for code using the `fs2`/`fs3`/`fs4` crates, all of which have a `FileExt` trait with methods called  `lock_shared`, `try_lock_shared`, and `unlock`.

This lint fires on the latest 1.83 beta release, so there might be a number of people who discover this once 1.83 stable goes out.

**Comment by joshtriplett** [comment]
While there are more features we may want to add to this in the future, the current state of this seems useful, and works. Stabilizing it would let people who encounter the warnings about a future conflict switch to the new API.

Shall we stabilize the current File locking APIs?

@rfcbot merge

**Comment by m-ou-se** [comment]
I'd like to see some documentation on the lock methods that make it clear that you don't need to manually unlock / that the lock is automatically unlocked when the File is dropped. Right now, that's only documented on the unlock method.

Other than that, the documentation could be made a lot less confusing by adding 'by another process' and 'by the same process' in a few places. Right now, the try methods say "Returns false if the file is locked.", but then go on to say it might deadlock if it's already locked. I assume the former should be "locked by another process" and the latter should be "locked by this process".

**Comment by m-ou-se** [comment]
The unlock method should document whether it's okay to call it if no locks are held.

If calling unlock() is only acceptable when a lock is actually held, this should probably be a Guard style of API. If it's always okay to call unlock(), the current design makes sense to me.

**Comment by m-ou-se** [comment]
I'm checking my box with the assumption that these are just small docs changes that we'll do during stabilization.

**Comment by joshtriplett** [comment]
I've submitted https://github.com/rust-lang/rust/pull/136288 which should address all the documentation requests.

@m-ou-se wrote:
> I'd like to see some documentation on the lock methods that make it clear that you don't need to manually unlock / that the lock is automatically unlocked when the File is dropped. Right now, that's only documented on the unlock method.

Done.

> Other than that, the documentation could be made a lot less confusing by adding 'by another process' and 'by the same process' in a few places. Right now, the try methods say "Returns false if the file is locked.", but then go on to say it might deadlock if it's already locked. I assume the former should be "locked by another process" and the latter should be "locked by this process".

"process" isn't the granularity here, but I've added clear distinctions about locks acquired via the same handle/descriptor (may deadlock) vs locks acquired via a different handle/descriptor (will block the blocking methods or make the `try_` methods return `Ok(false)`). I've also clarified the return value documentation to avoid that ambiguity.

> The unlock method should document whether it's okay to call it if no locks are held.

It's always safe to call (in the Rust sense). I've documented that it'll either return an error or return without doing anything. It'll never explode.

> If calling unlock() is only acceptable when a lock is actually held, this should probably be a Guard style of API. If it's always okay to call unlock(), the current design makes sense to me.

It's always OK (unlike a mutex or similar). It'll never explode.

**Comment by quininer** [comment]
I think this API needs more warnings for Windows, as `LockFile` might not be the right choice.
It has many inconsistencies with unix file lock, and using it blindly would be a mistake. like https://github.com/rust-lang/rust/issues/54118#issuecomment-1657196239 and [LockFileEx Remarks](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex#:~:text=If%20the%20locking%20process%20opens%20the%20file%20a%20second%20time,%20it%20cannot%20access%20the%20specified%20region) .

Finally, i used [share mode](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew#:~:text=[in]%20dwShareMode) instead of file lock, which is similar to the open options api mentioned by https://github.com/rust-lang/rust/issues/130994#issuecomment-2557089931.

**Comment by quininer** [comment]
@joshtriplett We already have [share_mode for Windows](https://doc.rust-lang.org/stable/std/os/windows/fs/trait.OpenOptionsExt.html#tymethod.share_mode), which is good. :D

> If the locking process opens the file a second time, it cannot access the specified region through this second handle until it unlocks the region.
> Locking a portion of a file for exclusive access denies all other processes both read and write access to the specified region of the file.

This means that the windows file lock is different from unix file lock in that it is not a advisory lock and it blocks file access. This is rather different from what std document implies

> Note, this is an advisory lock meant to interact with lock_shared, try_lock, try_lock_shared, and unlock.

I actually think the windows file lock is different enough from unix file lock that it should be in `FileExt` as well. Even if it isn't, there should be a more document warning.

**Comment by cberner** [comment]
@quininer the std documentation currently says `Its interactions with other methods, such as [`read`] and [`write`] are platform specific, and it may or may not cause non-lockholders to block.`

That's meant to say that the lock may be mandatory, depending on the platform (i.e. Windows). How would you make it more clear?

**Comment by joshtriplett** [comment]
I've clarified the documentation in https://github.com/rust-lang/rust/pull/136876 to address all of these issues.

**Comment by joshtriplett** [comment]
@ChrisDenton That's exactly what I've written in #136876.

**Comment by joshtriplett** [comment]
Making it just a Result would make it difficult to distinguish between failure to obtain the lock and another OS error. Code using try_lock may want to print a message and then try again blocking, for a failure to acquire the lock, but for an OS error they'll want to error out and exit.

Nested Result makes that case easy. Result of bool makes that case easy, though potentially error-prone until we have a way to mark the bool as must_use.

**Comment by BurntSushi** [comment]
I think `Result<(), TryLockError>` is probably the right call here too.

A nested `Result` is _weird_ IMO. I'm not above having used them before, but never in public APIs.

Having to do a bit of matching on the error type to distinguish the error cases for something like this seems very okay to me. Especially given that this API is probably rarely used, and so I think it doesn't need to be that ergonomic to use. But it shouldn't be bug-prone.

**Comment by moxian** [comment]
Since we are not sure the API is appropriate, it was suggested on [zullip](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/File.3A.3Atry_lock.20API.3A.20Result.3Cbool.3E/near/506823067) to revert the change in the short term while we figure this out. I've made https://github.com/rust-lang/rust/pull/138822

**Comment by jieyouxu** [comment]
Reopening because stabilization was reverted to reconsider the signature.

**Comment by cberner** [comment]
@BurntSushi `Result<(), TryLockError>` seems good to me too. Process-wise, anything I can do to help here? I can send a PR to make that change, but am not sure if I should open a new ACP first, since the [current one](https://github.com/rust-lang/libs-team/issues/412) says to use the `Result<bool>` style.

**Comment by BurntSushi** [comment]
I still think that's the best option, yes. I'm not sure if we have consensus yet though: https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/File.3A.3Atry_lock.20API.3A.20Result.3Cbool.3E

I don't think an ACP is needed for this sort of thing. ACPs are more like, "is the general direction of this API something that libs-api is receptive to." I think we have consensus that we want _an_ API here, but it sounds like we don't yet have consensus on its form. So the right place for iterating on that is this issue, Zulip and libs-api meetings.

**Comment by Amanieu** [comment]
We briefly discussed this in the @rust-lang/libs-api meeting today. Overall we felt that the best solution was indeed something along the lines of `Result<(), TryLockError>` where `TryLockError` is an enum with 2 variants.

**Comment by cberner** [comment]
@Amanieu sounds good. I will put together a PR to make that change. One other option came to mind though. Did you discuss having the signature be `Result<(), io::Error>` and returning `ErrorKind::WouldBlock` to indicate that lock acquisition failed?

On the one hand, it's nice because then `try_lock()?` can be used in a function that returns `io::Result`, but on the other hand it's less obvious from the signature how acquisition failure is handled.

**Comment by juntyr** [comment]
We could also add a From impl to convert the new error type to an io::Error

**Comment by cberner** [comment]
@juntyr oh ya, that gives the best of both worlds. I didn't realize that would automatically work with `?`. Thanks! I'll update the PR

**Comment by BurntSushi** [comment]
> [@juntyr](https://github.com/juntyr) oh ya, that gives the best of both worlds. I didn't realize that would automatically work with `?`. Thanks! I'll update the PR

Please keep that separate for now. It isn't clear to me that is a good idea, and it is an impl that can be added later after we gain experience with the API.

**Comment by cberner** [comment]
Sounds good. I reverted that part of the PR

**Comment by Amanieu** [comment]
> [@Amanieu](https://github.com/Amanieu?rgh-link-date=2025-04-03T23%3A45%3A20.000Z) sounds good. I will put together a PR to make that change. One other option came to mind though. Did you discuss having the signature be `Result<(), io::Error>` and returning `ErrorKind::WouldBlock` to indicate that lock acquisition failed?
> 
> On the one hand, it's nice because then `try_lock()?` can be used in a function that returns `io::Result`, but on the other hand it's less obvious from the signature how acquisition failure is handled.

This was discussed during the meeting and there was a slightly preference towards a custom error type since that makes it very obvious if the failure was due to a lock already existing or some other I/O error. Also note that currently Windows doesn't turn `ERROR_IO_PENDING` into `ErrorKind::WouldBlock` so that would have to be added.

**Comment by det** [comment]
Adding some personal experience. I find the new `try_lock` API less ergonomic than the previous version.

I can no longer write:
```
if file.try_lock()? {
    // Do this thing
} else {
    // Do this other thing.
}
```

I now have to write
```
match file.try_lock() {
    Ok(_) => {
        // Do this thing.
    }
    Err(TryLockError::WouldBlock) => {
        // Do this other thing.
    }
    Err(TryLockError::Error(err)) => return Err(err),
}
```

I have to do this in several places in my code base, so it makes sense for me (and probably others) to just add a helper function that is essentially the old API.

I think the fundamental problem is that the new API treats blocking as an error when the whole point of the API is to handle blocking gracefully.

If the goal is to make the return type more clear, another option is to replace the bool with an `enum TryLock { Locked, WouldBlock }`, but I think a bool is perfectly clear here already.

**Comment by cberner** [comment]
@joshtriplett the `try_lock()` and `try_lock_shared()` APIs have been updated. I just wanted to check in on the next steps for this feature. We already had a FCP, but then stabilization was reverted to make that change. Can we stabilize it again now that those have been updated, or should it go through a second FCP?

**Comment by det** [comment]
@cberner thanks, I didn't realize this was matching another API. I guess that makes sense then.

**Comment by joshtriplett** [comment]
Nominating for subsequent discussion.

**Comment by joshtriplett** [comment]
Speaking for myself:

I think the new API is a usability downgrade as well.

I think we may want to consider a dedicated two-value enum with `#[must_use]`, which would solve this problem. People can compare to the desired value with `==`, or we could add helper functions to that enum that return `bool`.

**Comment by joshtriplett** [comment]
@rfcbot concern api-is-unergonomic

**Comment by decathorpe** [comment]
> Treating try_lock failure as an error makes that pattern much more annoying to write.

Wouldn't this also apply to the existing `Mutex::try_lock`? It would be really awkward if there were two `try_lock` APIs in std that did basically the same thing but with different error handling ...

**Comment by joshtriplett** [comment]
We talked about this *at length* in today's @rust-lang/libs-api meeting.

We went back and forth about several different API designs. @Amanieu also looked extensively at code search results, and found that a *huge* number of users of `fs4` get this wrong, either by using `?` and ignoring the bool, or by looking at `.is_err()` or `.is_ok()` (and thus ignoring the bool). And even the ones that use `match` sometimes get this wrong, by looking for a would-block error in the `Err` case even though that can't actually happen.

We considered the possibility of `Result<WouldBlock, io::Error>`, where `WouldBlock` (name to be bikeshed) would be a two-value enum that has `#[must_use]`. However, several people felt that this would be *differently* error-prone, because people might check `.is_err()` or `.is_ok()`, which would still ignore the `WouldBlock`. @Amanieu advocated for the type signature to have lock acquisition failure be in the `Err` case, not the `Ok` case, for this reason.

Ultimately, the only API we could come to consensus on was `io::Result<()>`, with lock acquisition failure being an IO error with a `WouldBlock` error kind. That will avoid all of the common failure modes observed in code searches.

There are two interesting ways people want to use `try_lock`.

1) Bail on *any* failure, including lock acquisition failure; with the new API, that would be `file.try_lock()?;`.

2) Handle lock acquisition failure specifically. This new API does not make that case particularly easy, though it's not any *worse* than the match required by the current nightly API. Or, if you're willing to treat any failure as lock acquisition failure, you *can* use `if file.try_lock().is_err()`.

Given all of the above, and with the @rust-lang/libs-api team's apologies for the churn: we'd like to change the API to return `io::Result<()>`, with lock acquisition failure represented as an IO error with a `WouldBlock` error kind.

**Comment by dpc** [comment]
> and found that a huge number of users of fs4 get this wrong, either by using ? and ignoring the bool,

I'm myself surprised at this discovery. I thought that people will sometimes do that, but I thought the API will be just *mildly*-error prone. Well, great job actually checking.

> There are two interesting ways people want to use try_lock.
>  
> * Bail on any failure, including lock acquisition failure; with the new API, that would be file.try_lock()?;.

What's the rationale for this? It seems functionally it is never a good idea as the would-block and io-error (wrong permissions, etc.) have completely different consequences, and treating them as such seem to only result in higher level issues.

> we'd like to change the API to return `io::Result<()>`, with lock acquisition failure represented as an IO error with a `WouldBlock` error kind.

Personally, I'm OK with this. At least not acquiring the lock is an `Err`, so handling it the most naive way avoids the most dreadful scenario: running code that assumes it has a lock, while actually not having it. Comparing with `WouldBlock` seems self-explanatory for code reviews and such. And it matches the lower level semantics, so I there's also that.

**Comment by Amanieu** [comment]
> What's the rationale for this? It seems functionally it is never a good idea as the would-block and io-error (wrong permissions, etc.) have completely different consequences, and treating them as such seem to only result in higher level issues. Not a tragic mistake, but still probably always incorrect in theoretical sense.

There's quite a lot of code that treats failure to acquire a lock as a fatal error, usually CLI tools that would just report an error to the user.

**Comment by cberner** [comment]
> Given all of the above, and with the @rust-lang/libs-api team's apologies for the churn: we'd like to change the API to return io::Result<()>, with lock acquisition failure represented as an IO error with a WouldBlock error kind.

No worries, thanks for taking the time to discuss it! I've submitted https://github.com/rust-lang/rust/pull/140718 which changes the API to return `io::Result<()>`

**Comment by BurntSushi** [comment]
@Amanieu @joshtriplett Wait, what happened to the custom error type? That seems like it has all the benefits of `io::Result<()>` _plus_ the benefit of being a little more explicit. And we could make the `file.try_lock()?` call work in a function that returns `io::Result<T>` with a `From` impl.

**Comment by dpc** [comment]
> There's quite a lot of code that treats failure to acquire a lock as a fatal error, usually CLI tools that would just report an error to the user.

If they use a lock ... don't they want to wait for it to be released to acquire it? Or are these programs that do not expect to be ran in parallel and use the lock just as a safeguard mechanism?

**Comment by joshtriplett** [comment]
@BurntSushi In the last libs-api meeting, we felt that most people wouldn't be likely to care about the distinction, as long as the `Err` case was "didn't acquire the lock" (so that people using `.is_err()` or `.is_ok()` would get a correct answer). We also thought that if we *did* use a separate enum, we'd ultimately want a `From` impl to convert it to an `io::Error` or similar. So, given that, just using an `io::Error` to begin with seemed reasonable.

Do you have a use case for which a separate enum would be preferable?

**Comment by BurntSushi** [comment]
The use case is your [second point](https://github.com/rust-lang/rust/issues/130994#issuecomment-2855657455) in your previous comment. Trying to distinguish between the different failure cases in an `io::Error` is a pain. So having those cases explicitly delineated in an enum seems like a better API. With the `From` impl, you still achieve your first point.

**Comment by cberner** [comment]
I have a slight preference for keeping `TryLockError::WouldBlock`, but I don't have a strong opinion either way. I'll put together a PR to add the `From` impl for the current API, but lemme know if you all decide you'd rather go with returning `io::Error`

Here's [an example](https://github.com/cberner/redb/commit/2fc511974a626328e27d7bc48032141cdbfc16a9#diff-4f840d49c0d308e2761f600841bf21147eb9c63bdf05112ec66c83b10d21945aR35-R47) of how I use the current API, and you can see that I already handle multiple special error cases:
* lock not acquired because it's already held by another process
* lock not acquired because filesystem does not support locking
* lock not acquired due to some other I/O error

**Comment by Amanieu** [comment]
I personally have a very slight preference towards keeping `TryLockError` but would be happy with either choice. Given the shape of the discussion in the last few comments, I think we should just merge  #141312 and close #140718.

**Comment by fishface60** [comment]
Having banged my head against the Java file lock APIs recently I'm glad Rust hasn't decided to use the fcntl `F_SETLK` interfaces just because they include file range locking. It's been a lot of work to work around the limitations caused by that.

I like that this implementation documents what the platform-specific behaviours are by platform. It was irritating that the Java implementation documented that "some platforms may convert a request for a shared lock into an exclusive lock" and then I could find no trace of which one.

To its credit however, it gave recommendations for how to use the interfaces in a way that is safe cross-platform. I'm not sure how obvious it is from the Rust documentation that you should treat all locks as advisory and open files for write if you take an exclusive lock or read for a shared lock.

I think it may be important to document the platform-specific lock inheritance behaviour. It wasn't hugely relevant for Java since all implementations' locks were per-process, but with flock the lock will be inherited if the file descriptor is and with LockFileEx subprocesses don't inherit the lock.

**Comment by cberner** [comment]
@fishface60 

> I'm not sure how obvious it is from the Rust documentation that you should treat all locks as advisory

I tried to make this clear in this paragraph of the docs, but feel free to suggest improvements!:
```
    /// This lock may be advisory or mandatory. This lock is meant to interact with [`lock`],
    /// [`try_lock`], [`lock_shared`], [`try_lock_shared`], and [`unlock`]. Its interactions with
    /// other methods, such as [`read`] and [`write`] are platform specific, and it may or may not
    /// cause non-lockholders to block.
```

> open files for write if you take an exclusive lock or read for a shared lock

As far as I can tell that is not required. The tests pass on Linux and Windows even if I change them to use just read permission when acquiring an exclusive lock, and the Windows [documentation says](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex): `The handle must have been created with either the GENERIC_READ or GENERIC_WRITE access right.`

**Comment by cberner** [comment]
@joshtriplett it has been 10 days since the FCP started, but it seems like @rfcbot is stuck. Maybe due to [this issue](https://github.com/rust-lang/rfcbot-rs/issues/302)

Alright if I submit the stabilization PR?

**Comment by cberner** [comment]
@m-ou-se it looks like you fixed this rfcbot issue the [last time it happened](https://github.com/rust-lang/rust/issues/87096#issuecomment-1028792980). This FCP should have completed a week ago. Could you help get the FCP unstuck?

**Comment by ChrisDenton** [comment]
I'm not rustbot but I can do my best impression...

The final comment period, with a disposition to **merge**, as per the [review above](https://github.com/rust-lang/rust/issues/130994#issuecomment-2614265215), is now **complete**.

As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.

This will be merged soon.

**Comment by Kobzol** [comment]
Just a heads-up: this broke the compilation of Cargo 1.24.0 in the rustc-perf benchmark suite (https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/Stable.20cargo.20benchmark.20broke.20with.20latest.20beta). The problem is that Cargo was using the `fs2` crate and its [FileExt](https://docs.rs/fs2/latest/fs2/trait.FileExt.html) trait, which defines `try_lock_shared() -> io::Result<()>`. The new result type breaks that, as the stdlib method now takes precedence from the extension trait. Maybe we should do a crater run, just in case?

**Comment by ChrisDenton** [comment]
This is in beta now, right? So presumably the beta crater run should show up any issues.

**Comment by Kobzol** [comment]
Yeah, it showed up in https://github.com/rust-lang/rust/issues/143347. I was just thinking that maybe a crater run that is specific to this change could be useful, to filter out the results more. But since we now have the results from the beta crater run, it's probably useless.
